src/main/java/io/lettuce/core/metrics/MicrometerCommandLatencyRecorder.java

```diff
@@ -0,0 +1,116 @@
+import io.micrometer.core.instrument.Timer;
+import io.netty.channel.local.LocalAddress;
+
+import java.net.SocketAddress;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * @author Steven Sheehy
+ */
+public class MicrometerCommandLatencyRecorder implements CommandLatencyRecorder {
+
+    static final String LABEL_COMMAND = "command";
+
+    static final String LABEL_LOCAL = "local";
+
+    static final String LABEL_REMOTE = "remote";
+
+    static final String METRIC_COMPLETION = "lettuce.command.completion";
+
+    static final String METRIC_FIRST_RESPONSE = "lettuce.command.firstresponse";
+
+    private final MeterRegistry meterRegistry;
+
+    private final MicrometerCommandLatencyCollectorOptions options;
+
+    private final Map<CommandLatencyId, Timer> completionTimers = new ConcurrentHashMap<>();
+
+    private final Map<CommandLatencyId, Timer> firstResponseTimers = new ConcurrentHashMap<>();
+
+    public MicrometerCommandLatencyRecorder(MeterRegistry meterRegistry, MicrometerCommandLatencyCollectorOptions options) {
+        this.meterRegistry = meterRegistry;
+        this.options = options;
+    }
+
+    @Override
+    public void recordCommandLatency(SocketAddress local, SocketAddress remote, ProtocolKeyword protocolKeyword,
+            long firstResponseLatency, long completionLatency) {
+
+        if (!isEnabled()) {
+            return;
+        }
+
+        CommandLatencyId commandLatencyId = createId(local, remote, protocolKeyword);
+        Timer firstResponseTimer = firstResponseTimers.computeIfAbsent(commandLatencyId, this::firstResponseTimer);
+        firstResponseTimer.record(firstResponseLatency, TimeUnit.NANOSECONDS);
+
+        Timer completionTimer = completionTimers.computeIfAbsent(commandLatencyId, this::completionTimer);
+        completionTimer.record(completionLatency, TimeUnit.NANOSECONDS);
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return options.isEnabled();
+    }
+
+    private CommandLatencyId createId(SocketAddress local, SocketAddress remote, ProtocolKeyword commandType) {
+        return CommandLatencyId.create(options.localDistinction() ? local : LocalAddress.ANY, remote, commandType);
+    }
+
+    protected Timer completionTimer(CommandLatencyId commandLatencyId) {
+        Timer.Builder timer = Timer.builder(METRIC_COMPLETION)
+                .description("Latency between command send and command completion (complete response received")
+                .tag(LABEL_COMMAND, commandLatencyId.commandType().name())
+                .tag(LABEL_LOCAL, commandLatencyId.localAddress().toString())
+                .tag(LABEL_REMOTE, commandLatencyId.remoteAddress().toString())
+                .tags(options.tags());
+
+        if (options.isHistogram()) {
+            timer.publishPercentileHistogram()
+                    .publishPercentiles(options.targetPercentiles())
```

Comment:
It doesn't make sense to publish client-side calculated percentiles if the metrics backend supports server-side calculation from percentile histograms - the same and more percentiles can be approximated server-side and can be aggregated across tags. The client-side percentile calculations make additional time series that aren't needed.

You can't know in lettuce whether a meter registry at runtime will support percentile histograms or not, which is why this configuration is usually left to the user or as runtime-configuration. The current setup isn't a big issue since the histogram option is disabled by default, and users can still configure the timer through Micrometer outside of the options here.

Reply:
So we can remove the calls to `publishPercentileHistogram` and `publishPercentiles`?